compositor_pipeline/src/pipeline/encoder/ffmpeg_h264.rs

```diff
@@ -0,0 +1,321 @@
+            EncoderPreset::Faster => "4",
+            EncoderPreset::Fast => "6",
+            EncoderPreset::Medium => "7",
+            EncoderPreset::Slow => "8",
+            EncoderPreset::Slower => "9",
+            EncoderPreset::Veryslow => "10",
+            EncoderPreset::Placebo => "11",
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Options {
+    pub preset: EncoderPreset,
+    pub resolution: Resolution,
+    pub output_id: OutputId,
+}
+
+enum Message {
+    Frame(Frame),
+    Stop,
+}
+
+pub struct LibavH264Encoder {
+    resolution: Resolution,
+    output_id: OutputId,
+    frame_sender: Sender<Message>,
+    encoder_thread: Option<std::thread::JoinHandle<()>>,
+}
+
+impl LibavH264Encoder {
+    pub fn new(
+        options: Options,
+    ) -> Result<(Self, Box<dyn Iterator<Item = EncodedChunk> + Send>), EncoderInitError> {
+        let (frame_sender, frame_receiver) = crossbeam_channel::unbounded();
+        let (packet_sender, packet_receiver) = crossbeam_channel::unbounded();
+        let (result_sender, result_receiver) = crossbeam_channel::bounded(0);
+
+        let options_clone = options.clone();
+
+        let encoder_thread = std::thread::Builder::new()
+            .name(format!("Encoder thread for output {}", options.output_id))
+            .spawn(move || {
+                match Self::encoder_thread(
+                    options_clone,
+                    frame_receiver,
+                    packet_sender,
+                    &result_sender,
+                ) {
+                    Ok(_) => log::debug!("Encoder thread exited normally."),
+                    Err(e) => result_sender.send(Err(e)).unwrap(),
+                }
+            })
+            .unwrap();
+
+        result_receiver.recv().unwrap()?;
+
+        Ok((
+            Self {
+                encoder_thread: Some(encoder_thread),
+                frame_sender,
+                output_id: options.output_id,
+                resolution: options.resolution,
+            },
+            Box::new(packet_receiver.into_iter()),
+        ))
+    }
+
+    pub fn send_frame(&self, frame: Frame) {
+        match self.frame_sender.send(Message::Frame(frame)) {
+            Ok(_) => {}
+            Err(e) => {
+                error!("[output {}] Failed to send frame to encoder: {e}. It's possible the encoder thread has crashed.", self.output_id);
+            }
+        }
+    }
+
+    pub fn resolution(&self) -> Resolution {
+        self.resolution
+    }
+
+    fn encoder_thread(
+        options: Options,
+        frame_receiver: Receiver<Message>,
+        packet_sender: Sender<EncodedChunk>,
+        result_sender: &Sender<Result<(), EncoderInitError>>,
+    ) -> Result<(), EncoderInitError> {
+        let codec = ffmpeg_next::codec::encoder::find(Id::H264).ok_or(EncoderInitError::NoCodec)?;
+
+        let mut encoder = Context::new().encoder().video()?;
+
+        let pts_unit_secs = Rational::new(1, 90000);
```

Comment:
Where does this 90000 come from? I'm pretty sure, that this is the RTP clock rate.

Using `i64` as pts in `EncodedChunk` breaks the separation between input-decoder and encoder-output. Converting from/to duration depends on the RTP clock rate, which not only won't have any sense when other input/output sources will be added, but also is completely different for other RTP payload types (yeah, for video 90kHz is used most frequently, but audio can use 48kHz).

Reply:
I thought about this problem and decided this way of solving it was a bit better, but looking at it today I think I have a different opinion now. Maybe we can talk about this live?

Reply:
Sure